Mathlib/Tactic/CheckInstances.lean

```diff
@@ -0,0 +1,88 @@
+/-
+Copyright (c) 2023 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import Mathlib.Lean.Expr.Traverse
+import Mathlib.Control.Basic
+
+/-!
+# The `check_instances` tactic
+
+Resynthesizes all typeclasses in the goal,
+and reports for each whether this succeeds,
+and if so how closely it matches the original.
+-/
+
+set_option autoImplicit true
+
+open Lean Elab Meta Tactic
+namespace Lean.Expr
+
+/--
+Visit every typeclass instance in the `Expr`, calling a function `f` at each.
+-/
+def visitInstances (f : α → Expr → MetaM α) (init : α) (e : Expr) : MetaM α := do
+  e.traverse (init := init) fun a i => do
+    let ty ← inferType i
+    if ¬ ty.isForall && (← isClass? ty).isSome then
+       f a i
+    else
+      pure a
```

Comment:
Hmm, `traverse` isn't appropriate here because it'll fail to handle binders correctly (there will be free bvars).

Maybe the cleanest/quickest solution would be to use `Lean.Core.transform` with a custom monad to collect resynth results.

Reply:
Here's a version that uses `Lean.Meta.transform`. It also does deduplication of the results. There are some tests at the end that show it working under binders.
```lean
import Mathlib.Control.Basic

/-!
# The `check_instances` tactic

Resynthesizes all typeclasses in the goal,
and reports for each whether this succeeds,
and if so how closely it matches the original.
-/

set_option autoImplicit true

open Lean Elab Meta Tactic

namespace Mathlib.Tactic.CheckInstances

/-- The state for collecting instance problems from an expression.
The metavariables are assigned to the instance that appeared in the expression. -/
structure State where
  instances : Array MVarId := #[]

/-- Monad for collecting instance problems from an expression. -/
abbrev CheckInstanceM := StateT State MetaM

/-- Inserts a goal if it's not equal to other goals already present.
We say two goals are equal if
(1) their targets are syntatically equal,
(2) their values are syntactically equal, and
(3) they are defeq at reducible transparancy.
Returns the (possibly) updated state along with either the pre-existing goal or the new one. -/
def State.insert (s : State) (g : MVarId) : MetaM (State × MVarId) := do
  for g' in s.instances do
    if (← g.getType >>= instantiateMVars) == (← g'.getType >>= instantiateMVars) then
      if (← instantiateMVars (.mvar g)) == (← instantiateMVars (.mvar g')) then
        if ← withReducible <| withNewMCtxDepth <| isDefEqGuarded (.mvar g) (.mvar g') then
          return (s, g')
  return ({s with instances := s.instances.push g}, g)

/--
Visit every typeclass instance in the `Expr`, collecting them.
Returns an expression with the instances replaced with metavariables
assigned to the value of the instance, deduplicated.
-/
def collectInstances (e : Expr) : CheckInstanceM Expr := do
  Meta.transform (← instantiateMVars e)
    (pre := fun e' => do
      let ty ← inferType e'
      if (← isClass? ty).isSome then
        let g ← mkFreshExprMVar ty
        g.mvarId!.assign e'
        let s ← get
        let (s, g') ← s.insert g.mvarId!
        set s
        return .done (.mvar g')
      else
        return .continue)

/-- Collect all the instances appearing in `e`.
The are represented as metavariables assigned to the instance.
The purpose of this is to record the local context of the instance. -/
def runCollectInstances (e : Expr) : MetaM (Array MVarId) := do
  let (_, s) ← (collectInstances e).run {}
  return s.instances

/--
Result type for `checkInstance`.
Records whether an instance can be resynthesized,
and if so how closely it matches the original.
-/
inductive CheckInstance
  | failed
  | not_defeq (i : Expr)
  | defeq (i : Expr)
  | defeq_reducible (i : Expr)
  | exact (i : Expr)

/--
Attempt to resynthesize a typeclass, reporting whether this is possible,
and if so how closely it matches the original.
-/
def checkInstance (g : MVarId) : MetaM CheckInstance := g.withContext do
  let e ← instantiateMVars (.mvar g)
  if let .some t ← try? (do synthInstance (← g.getType)) then
    let t ← instantiateMVars t
    if e == t then
      return .exact t
    if ← withReducibleAndInstances <| isDefEq e t then
      return .defeq_reducible t
    else if ← isDefEq e t then
      return .defeq t
    else
      return .not_defeq t
  else
    return .failed

/--
Find all typeclass instances in an expression, and try resynthesizing them,
reporting whether this is possible and if so how closely it matches the original.
-/
def checkInstances (e : Expr) : MetaM (Array (MVarId × CheckInstance)) := do
  let insts ← runCollectInstances e
  insts.foldlM (init := #[]) fun a i => return a.push ⟨i, ← checkInstance i⟩

/--
A tactic for resynthesizing all typeclass instances in the goal,
reporting if they can be resynthesized, and if so how closely they match the original terms.
-/
elab "check_instances" : tactic => withMainContext do
  let results ← checkInstances (← getMainTarget)
  for ⟨i, r⟩ in results do
    i.withContext do
      let e ← instantiateMVars (.mvar i)
      match r with
      | .failed =>
          logInfo m!"💥: failed to resynthesize{indentD i}\nInstance in expression is{indentD e}"
      | .not_defeq e' =>
          logInfo m!"❌: resynthesized{indentD i}\nand got{indentD e'}\n{
            ""}which is not defeq to{indentD e}"
      | .defeq e' =>
          logInfo m!"🟡: resynthesized{indentD i}\nand got{indentD e'}\n{
            ""}which is defeq to{indentD e}"
      | .defeq_reducible e' =>
          logInfo m!"✅: resynthesized{indentD i}\nand got{indentD e'}\n{
              ""}which is defeq up to reducible transparancy to{indentD e}"
      | .exact _ =>
          logInfo m!"✅: resynthesized{indentD i}\nand got the same instance{indentD e}"

example : ∀ (α : Type) [DecidableEq α] (x y : α), if x = y then True else True := by
  check_instances
  intros; split <;> trivial

example : ∀ (x y : Nat), if x = y then True else True := by
  check_instances
  intros; split <;> trivial

-- With Mathlib.Tactic.Classical
example : ∀ (α : Type) [DecidableEq α] (x y : α), if x = y then True else True := by
  intros
  classical!
  check_instances -- resynth failure
  intros; split <;> trivial
```

Reply:
I made a prototype of a tactic for resynthesizing instances in the local context at #6930, and it can give a lot of the same information as `check_instances`. I think all that's missing is having it give finer-grained information about resynthesizing at different transparency levels.